### PR TITLE
feat: trying to make the receive caas file faster

### PR DESCRIPTION
**Pull request by SamRobinson75684**
## Type of changes

- [ ] Refactoring (non-breaking change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)

## Checklist

- [ ] I am familiar with the [contributing guidelines](../docs/CONTRIBUTING.md)
- [ ] I have followed the code style of the project
- [ ] I have added tests to cover my changes
- [ ] I have updated the documentation accordingly
- [ ] This PR is a result of pair or mob programming

---

## Sensitive Information Declaration

To ensure the utmost confidentiality and protect your and others privacy, we kindly ask you to NOT including [PII (Personal Identifiable Information) / PID (Personal Identifiable Data)](https://digital.nhs.uk/data-and-information/keeping-data-safe-and-benefitting-the-public) or any other sensitive data in this PR (Pull Request) and the codebase changes. We will remove any PR that do contain any sensitive information. We really appreciate your cooperation in this matter.

- [ ] I confirm that neither PII/PID nor sensitive data are included in this PR and the codebase changes.

### CODE REVIEW

File: application/CohortManager/src/Functions/CaasIntegration/receiveCaasFile/receiveCaasFile.cs [hunk 1]
```diff
@@ -72,54 +77,49 @@ public async Task Run([BlobTrigger("inbound/{name}", Connection = "caasfolder_ST
...
-                        if (!ValidationHelper.ValidateNHSNumber(participant.NhsNumber))
-                        {
-                            await _exceptionHandler.CreateSystemExceptionLog(new Exception($"Invalid NHS Number was passed at data row {rowNumber}"), participant, name);
-                            err++;
-                            return; // skip current participant
-                        }
+                    var listOfAllValues = values.ToList();
+                    var countOfRecords = values.Length;
 
+                    if (countOfRecords > 3)
```
Comment (MWClayson-NHS) [inline application/CohortManager/src/Functions/CaasIntegration/receiveCaasFile/receiveCaasFile.cs:98]: Please move these static numbers values out to environments variables or constants

File: application/CohortManager/src/Functions/CaasIntegration/receiveCaasFile/receiveCaasFile.cs [hunk 2]
```diff
@@ -136,6 +136,91 @@ await Parallel.ForEachAsync(values, async (rec, cancellationToken) =>
...
+        };
+
+        switch (participant.RecordType?.Trim())
+        {
+            case Actions.New:
+                //  we do this check in here because we can't do it in AddBatchToQueue with the rest of the calls
+                if (await _checkDemographic.PostDemographicDataAsync(basicParticipantCsvRecord.participant, Environment.GetEnvironmentVariable("DemographicURI")))
+                {
+                    // we need to wait on this dark to completed here
+                    await Task.CompletedTask;
```
Comment (MWClayson-NHS) [inline application/CohortManager/src/Functions/CaasIntegration/receiveCaasFile/receiveCaasFile.cs:203]: I don't think this line is needed any more

File: application/CohortManager/src/Functions/Shared/Common/AzureQueueStorageHelper.cs [hunk 3]
```diff
@@ -10,6 +11,8 @@ namespace Common;
 public class AzureQueueStorageHelper : IAzureQueueStorageHelper
```
Comment (MWClayson-NHS) [inline application/CohortManager/src/Functions/Shared/Common/AzureQueueStorageHelper.cs:11]: This is not generic enough for a helper class, there are alot of variables that are very specific one one process and should be handled outside of this.

File: application/CohortManager/src/Functions/Shared/Common/ProcessCaasFile.cs [hunk 4]
```diff
@@ -5,104 +5,74 @@ namespace Common;
...
-                await RemoveParticipant(basicParticipantCsvRecord, participant, filename);
-                break;
-            default:
-                await CreateError(participant, filename);
-                err++;
-                break;
-        }
-        _logger.LogInformation("There are {add} Additions. There are {upd} Updates. There are {del} Deletions. There are {err} Errors.", add, upd, del, err);
-
+        return await _azureQueueStorageHelper.CreateAddQueue();
```
Comment (MWClayson-NHS) [inline application/CohortManager/src/Functions/Shared/Common/ProcessCaasFile.cs:34]: making sure infra is available / created  should be done in the program.cs not in the execution of a function

File: application/CohortManager/src/Functions/CaasIntegration/receiveCaasFile/receiveCaasFile.cs
Comment (MWClayson-NHS) [inline application/CohortManager/src/Functions/CaasIntegration/receiveCaasFile/receiveCaasFile.cs:1]: This file has alot of logic in it, can this be broken down in to another class with methods names that make it clear what each step is doing?

**Review by MWClayson-NHS — COMMENTED** [review]
Please can you look at the issue raised by Sonar aswell?

File: application/CohortManager/src/Functions/CaasIntegration/receiveCaasFile/ProcessFileClasses/ProcessCaasFile.cs
Comment (MWClayson-NHS) [inline application/CohortManager/src/Functions/CaasIntegration/receiveCaasFile/ProcessFileClasses/ProcessCaasFile.cs:1]: Please can you order the methods in this file following the standard?

File: application/CohortManager/src/Functions/CaasIntegration/receiveCaasFile/receiveCaasFile.cs [hunk 5]
```diff
@@ -64,62 +72,47 @@ public async Task Run([BlobTrigger("inbound/{name}", Connection = "caasfolder_ST
...
+                    int.TryParse(Environment.GetEnvironmentVariable("batchDivisionFactor"), out var batchDivisionFactor);
+                    if (countOfRecords > recordThresholdForBatching)
+                    {
+                        //split list of all into N amount of chunks to be processed as batches
+                        var chunkSize = countOfRecords / batchDivisionFactor;
+                        var chunks = listOfAllValues.Chunk(chunkSize).ToList();
 
-                        if (!_receiveCaasFileHelper.validateDateTimes(participant))
+                        var allTasks = new List<Task>();
+                        foreach (var chunk in chunks)
```
Comment (MWClayson-NHS) [inline application/CohortManager/src/Functions/CaasIntegration/receiveCaasFile/receiveCaasFile.cs:98]: Can this get moved out a bit there is alot of nesting

**Review by MWClayson-NHS — DISMISSED** [review]

**Review by Warren-Pitterson — APPROVED** [review]

File: application/CohortManager/src/Functions/CaasIntegration/receiveCaasFile/ProcessFileClasses/AddBatchToQueue.cs [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+    public AddBatchToQueue(ILogger<AddBatchToQueue> logger)
+    {
+        storageConnectionString = Environment.GetEnvironmentVariable("AzureWebJobsStorage") ?? "";
+        _AddQueueClient = new QueueClient(storageConnectionString, Environment.GetEnvironmentVariable("AddQueueName"));
+
+        _AddQueueClient.CreateIfNotExists();
+        _logger = logger;
+    }
+
```
Comment (Warren-Pitterson) [inline application/CohortManager/src/Functions/CaasIntegration/receiveCaasFile/ProcessFileClasses/AddBatchToQueue.cs:25]: whitespace please remove

**Review by MWClayson-NHS — APPROVED** [review]

**Review by Warren-Pitterson — APPROVED** [review]

**Review by MWClayson-NHS — APPROVED** [review]

**Pull request merged**